contracts/rust/src/light_client.rs

```diff
@@ -0,0 +1,405 @@
+                        state_msg,
+                        &mut self.rng,
+                    )
+                } else {
+                    Ok(Signature::<EdwardsConfig>::default())
+                }
+            })
+            .collect::<Result<Vec<_>, _>>()
+            .unwrap();
+
+        let srs = {
+            // load SRS from Aztec's ceremony
+            let srs = crs::aztec20::kzg10_setup(2u64.pow(16) as usize + 2)
+                .expect("Aztec SRS fail to load");
+            // convert to Jellyfish type
+            // TODO: (alex) use constructor instead https://github.com/EspressoSystems/jellyfish/issues/440
+            UnivariateUniversalParams {
+                powers_of_g: srs.powers_of_g,
+                h: srs.h,
+                beta_h: srs.beta_h,
+                powers_of_h: vec![srs.h, srs.beta_h],
+            }
+        };
+        let (pk, _) = hotshot_state_prover::preprocess::<STAKE_TABLE_CAPACITY>(&srs)
+            .expect("Fail to preprocess state prover circuit");
+        let (proof, pi) =
+            hotshot_state_prover::generate_state_update_proof::<_, _, _, _, STAKE_TABLE_CAPACITY>(
+                &mut self.rng,
+                &pk,
+                &self.st,
+                &bit_vec,
+                &sigs,
+                &self.state,
+                &self.threshold,
+            )
+            .expect("Fail to generate state proof");
+        (pi, proof)
+    }
+
+    /// Returns the `LightClientState` for solidity
+    pub fn get_state(&self) -> ParsedLightClientState {
+        // The ugly conversion due to slight difference of `LightClientState` in solidity containing `threshold`
+        let pi = vec![
+            u256_to_field(self.threshold),
+            F::from(self.state.view_number as u64),
+            F::from(self.state.block_height as u64),
+            self.state.block_comm_root,
+            self.state.fee_ledger_comm,
+            self.state.stake_table_comm.0,
+            self.state.stake_table_comm.1,
+            self.state.stake_table_comm.2,
+        ];
+        let pi: PublicInput<F> = pi.into();
+        pi.into()
+    }
+
+    /// Returns the (bytes32 votingStakeTableComm, bytes32 frozenStakeTableComm) used in contract
+    pub fn get_stake_table_comms(&self) -> (H256, H256) {
+        let (bls_key_comm, schnorr_key_comm, amount_comm) =
+            self.st.commitment(SnapshotVersion::EpochStart).unwrap();
+        let frozen_st_comm = utils::keccak256(
+            abi::encode_packed(&[
+                Token::Uint(field_to_u256(bls_key_comm)),
+                Token::Uint(field_to_u256(schnorr_key_comm)),
+                Token::Uint(field_to_u256(amount_comm)),
+            ])
+            .unwrap(),
+        );
+
+        let (bls_key_comm, schnorr_key_comm, amount_comm) =
+            self.st.commitment(SnapshotVersion::LastEpochStart).unwrap();
+        let voting_st_comm = utils::keccak256(
```

Comment:
Use `computeStakeTableComm` instead?

Reply:
are you suggesting we use the API from rust's contract-binding? I think it's actually not bad to do it here, we will be having the same code to compute stake table commitments in sequencer code anyway.

Reply:
Sorry, I thought I had deleted this comment, but it is still there. No need t change anything.